test/test_transforms_v2_refactored.py

```diff
@@ -0,0 +1,679 @@
+    if not isinstance(input, datapoints.Image):
+        return
+
+    dispatcher_scripted = _script(dispatcher)
+    with ignore_jit_no_profile_information_warning():
+        dispatcher_scripted(input.as_subclass(torch.Tensor), *args, **kwargs)
+
+
+def _check_dispatcher_dispatch_simple_tensor(dispatcher, kernel, input, *args, **kwargs):
+    """Checks if the dispatcher correctly dispatches simple tensors to the ``*_image_tensor`` kernel and the input type
+    is preserved in doing so.
+    """
+    if type(input) is not torch.Tensor:
+        return
+
+    with mock.patch(f"{dispatcher.__module__}.{kernel.__name__}", wraps=kernel) as spy:
+        output = dispatcher(input, *args, **kwargs)
+
+        spy.assert_called_once()
+
+    # We cannot use `isinstance` here since all datapoints are instances of `torch.Tensor` as well
+    assert type(output) is torch.Tensor
+
+
+def _check_dispatcher_dispatch_pil(dispatcher, kernel, input, *args, **kwargs):
+    """Checks if the dispatcher correctly dispatches PIL images to the ``*_image_pil`` kernel  and the input type
+    is preserved in doing so.
+    """
+    if not isinstance(input, PIL.Image.Image):
+        return
+
+    with mock.patch(f"{dispatcher.__module__}.{kernel.__name__}", wraps=kernel) as spy:
+        output = dispatcher(input, *args, **kwargs)
+
+        spy.assert_called_once()
+
+    assert isinstance(output, PIL.Image.Image)
+
+
+def _check_dispatcher_dispatch_datapoint(dispatcher, kernel, input, *args, **kwargs):
+    """Checks if the dispatcher ultimately correctly dispatches datapoints to corresponding kernel and the input type
+    is preserved in doing so. For bounding boxes also checks that the format is preserved.
+    """
+    if not isinstance(input, datapoints._datapoint.Datapoint):
+        return
+
+    # Due to our complex dispatch architecture for datapoints, we cannot spy on the kernel directly,
+    # but rather have to patch the `Datapoint.__F` attribute to contain the spied on kernel.
+    spy = mock.MagicMock(wraps=kernel)
+    with mock.patch.object(F, kernel.__name__, spy):
+        # Due to Python's name mangling, the `Datapoint.__F` attribute is only accessible from inside the class.
+        # Since that is not the case here, we need to prefix f"_{cls.__name__}"
+        # See https://docs.python.org/3/tutorial/classes.html#private-variables for details
+        with mock.patch.object(datapoints._datapoint.Datapoint, "_Datapoint__F", new=F):
+            output = dispatcher(input, *args, **kwargs)
+
+    spy.assert_called_once()
+    assert isinstance(output, type(input))
+
+    if isinstance(input, datapoints.BoundingBox):
+        assert output.format == input.format
+
+
+def check_dispatcher(
+    dispatcher,
+    kernel,
+    input,
+    *args,
+    check_scripted_smoke=True,
+    check_dispatch_simple_tensor=True,
+    check_dispatch_pil=True,
+    check_dispatch_datapoint=True,
```

Comment:
Do we already know which of the transforms  are going to set one of these to False?
I'm not sure about the rest but it sounds at least that `check_dispatch_simple_tensor` and `check_dispatch_datapoint` should always be True?

In many places we have parameters that we thought could be useful eventually, but ended up never being used and clutter the API. It' best to be conservative and only expose what we need when we need it (unless we're 100% sure we'll need it very very soon)

Reply:
> Do we already know which of the transforms are going to set one of these to False?

Yes, there are a few. You can see them by looking at the configs in [`transforms_v2_dispatcher_infos.py`](https://github.com/pytorch/vision/blob/main/test/transforms_v2_dispatcher_infos.py) that use the following skip marker.

https://github.com/pytorch/vision/blob/fc83b280387d92fe6aea73f35c1ed4ad1bb739da/test/transforms_v2_dispatcher_infos.py#L99-L102

> I'm not sure about the rest but it sounds at least that `check_dispatch_simple_tensor` and `check_dispatch_datapoint` should always be True?

Indeed, the statement above only applies to datapoints. To the best of my knowledge we don't have a dispatcher for that we can't automatically check simple tensor and PIL dispatch.